doc/man/man8/slapacl.8
.TH SLAPACL 8C "RELEASEDATE" "OpenLDAP LDVERSION"
.\" Copyright 2004-2011 The OpenLDAP Foundation All Rights Reserved.
.\" Copying restrictions apply.  See COPYRIGHT/LICENSE.
.\" $OpenLDAP: pkg/ldap/doc/man/man8/slapacl.8,v 1.8.2.12 2010/04/13 20:22:45 kurt Exp $
.SH NAME
slapacl \- Check access to a list of attributes.
.SH SYNOPSIS
.B SBINDIR/slapacl
.BI \-b \ DN
[\c
.BI \-d \ debug-level\fR]
[\c
.BI \-D \ authcDN\ \fR|
.BI \-U \ authcID\fR]
[\c
.BI \-f \ slapd.conf\fR]
[\c
.BI \-F \ confdir\fR]
[\c
.BI \-o \ option\fR[ = value\fR]]
[\c
.BR \-u ]
[\c
.BR \-v ]
[\c
.BI \-X \ authzID\ \fR|
.BI "\-o \ authzDN=" DN\fR]
[\c
.IR attr [\fB/\fI access ][\fB:\fI value ]]\fR\ [...]
.LP
.SH DESCRIPTION
.LP
.B slapacl
is used to check the behavior of 
.BR slapd (8) 
by verifying access to directory data according to the access control list
directives defined in its configuration.
.
It opens the
.BR slapd.conf (5)
configuration file or the 
.BR slapd\-config (5) 
backend, reads in the  
.BR access / olcAccess
directives, and then parses the 
.B attr
list given on the command-line; if none is given, access to the
.B entry
pseudo-attribute is tested.
.LP
.SH OPTIONS
.TP
.BI \-b \ DN
specify the 
.I DN 
which access is requested to; the corresponding entry is fetched 
from the database, and thus it must exist.
The
.I DN
is also used to determine what rules apply; thus, it must be
in the naming context of a configured database.  See also
.BR \-u .
.TP
.BI \-d \ debug-level
enable debugging messages as defined by the specified
.IR debug-level ;
see
.BR slapd (8)
for details.
.TP
.BI \-D \ authcDN
specify a DN to be used as identity through the test session
when selecting appropriate
.B <by> 
clauses in access lists.
.TP
.BI \-f \ slapd.conf
specify an alternative
.BR slapd.conf (5)
file.
.TP
.BI \-F \ confdir
specify a config directory.
If both
.B \-f
and
.B \-F
are specified, the config file will be read and converted to
config directory format and written to the specified directory.
If neither option is specified, an attempt to read the
default config directory will be made before trying to use the default
config file. If a valid config directory exists then the
default config file is ignored.
.TP
.BI \-o \ option\fR[ = value\fR]
Specify an
.I option
with a(n optional)
.IR value .
Possible generic options/values are:
.LP
.nf
              syslog=<subsystems>  (see `\-s' in slapd(8))
              syslog\-level=<level> (see `\-S' in slapd(8))
              syslog\-user=<user>   (see `\-l' in slapd(8))

.fi
.RS
Possible options/values specific to
.B slapacl
are:
.RE
.nf

              authzDN
              domain
              peername
              sasl_ssf
              sockname
              sockurl
              ssf
              tls_ssf
              transport_ssf

.fi
.RS
See the related fields in
.BR slapd.access (5)
for details.
.RE
.TP
.BI \-u
do not fetch the entry from the database.
In this case, if the entry does not exist, a fake entry with the
.I DN
given with the
.B \-b
option is used, with no attributes.
As a consequence, those rules that depend on the contents 
of the target object will not behave as with the real object.
The
.I DN
given with the
.B \-b
option is still used to select what rules apply; thus, it must be
in the naming context of a configured database.
See also
.BR \-b .
.TP
.BI \-U \ authcID
specify an ID to be mapped to a 
.B DN 
as by means of 
.B authz\-regexp
or
.B authz\-rewrite
rules (see 
.BR slapd.conf (5)
for details); mutually exclusive with
.BR \-D .
.TP
.B \-v
enable verbose mode.
.TP
.BI \-X \ authzID
specify an authorization ID to be mapped to a
.B DN
as by means of
.B authz\-regexp
or
.B authz\-rewrite
rules (see
.BR slapd.conf (5)
for details); mutually exclusive with \fB\-o\fP \fBauthzDN=\fIDN\fR.
.SH EXAMPLES
The command
.LP
.nf
.ft tt
	SBINDIR/slapacl \-f ETCDIR/slapd.conf \-v \\
            \-U bjorn \-b "o=University of Michigan,c=US" \\
	    "o/read:University of Michigan"

.ft
.fi
tests whether the user
.I bjorn
can access the attribute 
.I o
of the entry
.I o=University of Michigan,c=US
at
.I read
level.
.SH "SEE ALSO"
.BR ldap (3),
.BR slapd (8),
.BR slaptest (8),
.BR slapauth (8)
.LP
"OpenLDAP Administrator's Guide" (http://www.OpenLDAP.org/doc/admin/)
.SH ACKNOWLEDGEMENTS
.so ../Project


CHANGES
OpenLDAP 2.0 Change Log

OpenLDAP 2.0.27 Release
	Build environment
		Fix getopt(3) detection

OpenLDAP 2.0.26 Release
	Fixed slapd ACL buffer size bug (ITS#1963)
	Fixed slapd ACL filters w/substrings bug (ITS#2023, ITS#2027)
	Fxied slapd directory string space bug (ITS#2072)
	Fixed slurpd reject file handling bugs (ITS#2007, ITS#2009)
	Fixed back-ldbm indirect IDL bug (ITS#2039)
	Fixed libldap buffer size bugs (ITS#1935, ITS#1936)
	Fixed libldap gethostbyname mutex init bug (ITS#2019)
	Fixed liblutil passwd bug (ITS#2055)
	Fixed sasl large reads and writes (ITS#1983, also ITS#910,ITS#1861)
	Updated back-ldbm reindexing - slapindex rebuilds dn2id index
	Build environment
		Require version 1.5 if Cyrus SASL is present
		Remove tempnam checking and replacment implementation

OpenLDAP 2.0.25 Release
	Fixed slapd extended op reqdata crash bug (ITS#1721)

OpenLDAP 2.0.24 Release
	Fixed slapd max incoming macro bug (ITS#1828)
	Fixed slapd acl group/dnaddr bug (ITS#1607)
	Fixed slapd connection management bug (ITS#1655)
	Fixed slapd port ntoh / addr init bugs (ITS#1615, ITS#1728)
	Fixed slapd PF_lOCAL typo (ITS#1660)
	Fixed slapd extended ops with NULL reqdata bug (ITS#1721)
	Fixed slapd replog logging without replica (ITS#1335)
	Fixed slapd NULL attrval indexing (ITS#1843)
	Fixed slurpd tls init bug (ITS#1613)
	Fixed back-ldbm dn normalization bug in onelevel searches (ITS#1654)
	Fixed back-ldbm modrdn root dn check (ITS#1761)
	Fixed back-ldbm idl overrun bug (ITS#1570)
	Fixed back-ldbm passwd textlen bug (ITS#1742)
	Fixed ber_bvstrdup() empty string bug (ITS#1662)
	Fixed ldapsearch filter bug (ITS#1649)
	Fixed maildap NULL domain bug (ITS#1022)
	Fixed libldap hostless URL bug (ITS#1697)
	Fixed liblutil /dev/urandom incomplete reads (ITS#1834)
	Fixed libldap_r threadpool free list (ITS#1839)
	Fixed libldap_r NT threads cond_broadcast (ITS#1865)
	Fixed sasl writes > 64K (ITS#1861)
	Fixed sasl session callback corruption (ITS#1799)
	Build environment
		Removed lint (ITS#1744)
	Documentation
		Updated release documents

OpenLDAP 2.0.23 Release
	Fixed slapd moddn acl bug (ITS#1562) (ITS#1582)
	Fixed slapd acl regex (to dn="") portability bug (ITS#1579)
	Disabled -lldap atexit() handler (ITS#1577)
	Build environment
		Updated config.guess/sub scripts (ITS#1567)
	Documentation
		Fixed a number of typos (ITS#1578)

OpenLDAP 2.0.22 Release
	Fixed slapd passwd modify referral bug
	Fixed back-ldbm index threading bug
	Fixed back-ldbm ordering presence index bug
	Fixed back-ldap modify bug (ITS#1547)
	Fixed -lldap TLS errno bug
	Fixed ud v2 kbind bug
	Updated slapd slurpd friendly code
	Build environment
		Fixed Mach cthreads detection (ITS#1554)
		Updated Berkeley DB 4 detection

OpenLDAP 2.0.21 Release
	Fix slapd empty string indexing/filtering handling (ITS#1507)
	Changed slapd ACL selfwrite replace logic (ITS#1530)
	Build environment
		Updated IPv6 configuration argument handling
		Fix back-shell tools make depend bug (ITS#1518)
		Removed extraneous files

OpenLDAP 2.0.20 Release
	Fixed slapd ACL empty replace bug (ITS#1530)
	Fixed slapd ACL peername/sockname exact match bug (ITS#1516)
	Fixed back-passwd db_config bug
	Fixed -lldap cache debug bug (ITS#1501)
	Fixed -lldap dnssrv free bug
	Fixed -lldap Start TLS first op bug
	Fixed slurpd TLS non-critical/critical bug (ITS#1527)
	Fixed slurpd SASL password bug (ITS#1528)
	Build environment
		Fixed repl_user build error (ITS#1503)
		Updated BerkeleyDB 4 support
		Updated schema files
		Updated Windows NT/2000 MSVC5 port

OpenLDAP 2.0.19 Release
	Fixed back-ldbm not filter indexing (ITS#1405)
	Fixed back-passwd e_ndn bug (ITS#1467)
	Fixed back-ldap undef ad bug (ITS#1367)
	Fixed slapd suffix "" bugs (ITS#1430)
	Fixed slapd zero length replace bug (ITS#1364)
	Fixed slapd teletex string printables bug (ITS#1419)
	Fixed slapd unallocated sockets bug (ITS#1415)
	Fixed slapd undefined present filter bug (ITS#1439)
	Fixed -lldap cacertfile bug (ITS#1406)
	Fixed -lldap ldap_url_seach_s bug (ITS#1429)
	Fixed -lldap unbind free bug (ITS#1442)
	Fixed -llber realloc bug (ITS#1410)
	Fixed ldappasswd/slappasswd strncmp bug (ITS#1497)
	Fixed various memory leaks
	Updated slapd emfile/enfile handling
	Build environment
		Fixed passwd-search port bug (ITS#1461)
		Fixed Kerberos build bug (ITS#1468)
		Updated CThread detection (ITS#1440)
		Updated MAXPATHLEN handling (ITS#1440)
		Fixed TLS build (ITS#1412)
		Added BerkeleyDB 4 support
		Added GNU ar support (ITS#1460)

OpenLDAP 2.0.18 Release
	Fixed -llber ber_bvecadd bug (ITS#1398)
	Fixed slap tools replica config bug (ITS#1385)
	Build environment
		Updated req_query detection

OpenLDAP 2.0.17 Release
	Fixed ldbm BerkeleyDB2 compat bug (ITS#1390)

OpenLDAP 2.0.16 Release
	Fixed slapd indexing bug (ITS#1349)
	Fixed slapd/slurpd maxargs bug (ITS#1343)
	Fixed slapd ManageDSAIT warning
	Fixed slapd no OID schema bug
	Fixed slurpd server down reject bug (ITS#1183)
	Fixed slurpd moddn newsup bug (ITS#1379)
	Fixed -llber ber_realloc bug (ITS#1346)
	Fixed -lldap multi-level referral chase bug (ITS#1346)
	Fixed -lldap high port bug (ITS#1375)
	Fixed CR/LF handling (ITS#1328)
	Updated slapd/-lldap getaddrinfo handling (ITS#1304)
	Updated -lldap_r pthread stack size
	Updated slapd acl assert checks
	Updated slurpd ManageDSAIT handling
	Updated slapd root/subschema DSE update error handling
	Removed -lldbm use of Berkeley CDB
	Updated -lldbm to use r/w lock with DB_THREAD
	Updated back-ldbm update attribute indices on delete (ITS#1359)
	Updated ud Kerberos ticket handling
	Added back-ldbm idl_check() code
	Build environment
		Updated ltconfig for MacOS X 10.1 (ITS#1361)
		Updated slapd.conf with ACL example
		Updated pthread detection
		Updated Kerberos detection (ITS#1372)
		Updated test005 /dev/null diff for Tru64 compat
		Updated test007 modrdn test
		Updated test008 read/write ratio
		Updated MSVC5 project files
	Documentation
		Update slapcat/add pages

OpenLDAP 2.0.15 Release
	Fixed -lldap TLS external handling

servers/slapd/slapacl.c
/* $OpenLDAP: pkg/ldap/servers/slapd/slapacl.c,v 1.24.2.11 2010/04/13 20:23:20 kurt Exp $ */
/* This work is part of OpenLDAP Software <http://www.openldap.org/>.
 *
 * Copyright 2004-2011 The OpenLDAP Foundation.
 * Portions Copyright 2004 Pierangelo Masarati.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted only as authorized by the OpenLDAP
 * Public License.
 *
 * A copy of this license is available in file LICENSE in the
 * top-level directory of the distribution or, alternatively, at
 * <http://www.OpenLDAP.org/license.html>.
 */
/* ACKNOWLEDGEMENTS:
 * This work was initially developed by Pierangelo Masarati for inclusion
 * in OpenLDAP Software.
 */

#include "portable.h"

#include <stdio.h>

#include <ac/stdlib.h>

#include <ac/ctype.h>
#include <ac/string.h>
#include <ac/socket.h>
#include <ac/unistd.h>

#include <lber.h>
#include <ldif.h>
#include <lutil.h>

#include "slapcommon.h"

static int
print_access(
	Operation		*op,
	Entry			*e,
	AttributeDescription	*desc,
	struct berval		*val,
	struct berval		*nval )
{
	int			rc;
	slap_mask_t		mask;
	char			accessmaskbuf[ACCESSMASK_MAXLEN];

	rc = access_allowed_mask( op, e, desc, nval, ACL_AUTH, NULL, &mask );

	fprintf( stderr, "%s%s%s: %s\n",
			desc->ad_cname.bv_val,
			( val && !BER_BVISNULL( val ) ) ? "=" : "",
			( val && !BER_BVISNULL( val ) ) ?
				( desc == slap_schema.si_ad_userPassword ?
					"****" : val->bv_val ) : "",
			accessmask2str( mask, accessmaskbuf, 1 ) );

	return rc;
}

int
slapacl( int argc, char **argv )
{
	int			rc = EXIT_SUCCESS;
	const char		*progname = "slapacl";
	Connection		conn = { 0 };
	Listener		listener;
	OperationBuffer	opbuf;
	Operation		*op = NULL;
	Entry			e = { 0 }, *ep = &e;
	char			*attr = NULL;
	int			doclose = 0;
	BackendDB		*bd;
	void			*thrctx;

	slap_tool_init( progname, SLAPACL, argc, argv );

	if ( !dryrun ) {
		int	i = 0;

		LDAP_STAILQ_FOREACH( bd, &backendDB, be_next ) {
			if ( bd != be && backend_startup( bd ) ) {
				fprintf( stderr, "backend_startup(#%d%s%s) failed\n",
						i,
						bd->be_suffix ? ": " : "",
						bd->be_suffix ? bd->be_suffix[0].bv_val : "" );
				rc = 1;
				goto destroy;
			}

			i++;
		}
	}

	argv = &argv[ optind ];
	argc -= optind;

	thrctx = ldap_pvt_thread_pool_context();
	connection_fake_init( &conn, &opbuf, thrctx );
	op = &opbuf.ob_op;
	op->o_tmpmemctx = NULL;

	conn.c_listener = &listener;
	conn.c_listener_url = listener_url;
	conn.c_peer_domain = peer_domain;
	conn.c_peer_name = peer_name;
	conn.c_sock_name = sock_name;
	op->o_ssf = ssf;
	op->o_transport_ssf = transport_ssf;
	op->o_tls_ssf = tls_ssf;
	op->o_sasl_ssf = sasl_ssf;

	if ( !BER_BVISNULL( &authcID ) ) {
		if ( !BER_BVISNULL( &authcDN ) ) {
			fprintf( stderr, "both authcID=\"%s\" "
					"and authcDN=\"%s\" provided\n",
					authcID.bv_val, authcDN.bv_val );
			rc = 1;
			goto destroy;
		}

		rc = slap_sasl_getdn( &conn, op, &authcID, NULL,
				&authcDN, SLAP_GETDN_AUTHCID );
		if ( rc != LDAP_SUCCESS ) {
			fprintf( stderr, "authcID: <%s> check failed %d (%s)\n",
					authcID.bv_val, rc,
					ldap_err2string( rc ) );
			rc = 1;
			goto destroy;
		}

	} else if ( !BER_BVISNULL( &authcDN ) ) {
		struct berval	ndn;

		rc = dnNormalize( 0, NULL, NULL, &authcDN, &ndn, NULL );
		if ( rc != LDAP_SUCCESS ) {
			fprintf( stderr, "autchDN=\"%s\" normalization failed %d (%s)\n",
					authcDN.bv_val, rc,
					ldap_err2string( rc ) );
			rc = 1;
			goto destroy;
		}
		ch_free( authcDN.bv_val );
		authcDN = ndn;
	}

	if ( !BER_BVISNULL( &authzID ) ) {
		if ( !BER_BVISNULL( &authzDN ) ) {
			fprintf( stderr, "both authzID=\"%s\" "
					"and authzDN=\"%s\" provided\n",
					authzID.bv_val, authzDN.bv_val );
			rc = 1;
			goto destroy;
		}

		rc = slap_sasl_getdn( &conn, op, &authzID, NULL,
				&authzDN, SLAP_GETDN_AUTHZID );
		if ( rc != LDAP_SUCCESS ) {
			fprintf( stderr, "authzID: <%s> check failed %d (%s)\n",
					authzID.bv_val, rc,
					ldap_err2string( rc ) );
			rc = 1;
			goto destroy;
		}

	} else if ( !BER_BVISNULL( &authzDN ) ) {
		struct berval	ndn;

		rc = dnNormalize( 0, NULL, NULL, &authzDN, &ndn, NULL );
		if ( rc != LDAP_SUCCESS ) {
			fprintf( stderr, "autchDN=\"%s\" normalization failed %d (%s)\n",
					authzDN.bv_val, rc,
					ldap_err2string( rc ) );
			rc = 1;
			goto destroy;
		}
		ch_free( authzDN.bv_val );
		authzDN = ndn;
	}


	if ( !BER_BVISNULL( &authcDN ) ) {
		fprintf( stderr, "authcDN: \"%s\"\n", authcDN.bv_val );
	}

	if ( !BER_BVISNULL( &authzDN ) ) {
		fprintf( stderr, "authzDN: \"%s\"\n", authzDN.bv_val );
	}

	if ( !BER_BVISNULL( &authzDN ) ) {
		op->o_dn = authzDN;
		op->o_ndn = authzDN;
		
		if ( !BER_BVISNULL( &authcDN ) ) {
			op->o_conn->c_dn = authcDN;
			op->o_conn->c_ndn = authcDN;

		} else {
			op->o_conn->c_dn = authzDN;
			op->o_conn->c_ndn = authzDN;
		}

	} else if ( !BER_BVISNULL( &authcDN ) ) {
		op->o_conn->c_dn = authcDN;
		op->o_conn->c_ndn = authcDN;
		op->o_dn = authcDN;
		op->o_ndn = authcDN;
	}

	assert( !BER_BVISNULL( &baseDN ) );
	rc = dnPrettyNormal( NULL, &baseDN, &e.e_name, &e.e_nname, NULL );
	if ( rc != LDAP_SUCCESS ) {
		fprintf( stderr, "base=\"%s\" normalization failed %d (%s)\n",
				baseDN.bv_val, rc,
				ldap_err2string( rc ) );
		rc = 1;
		goto destroy;
	}

	op->o_bd = be;
	if ( op->o_bd == NULL ) {
		/* NOTE: if no database could be found (e.g. because
		 * accessing the rootDSE or so), use the frontendDB
		 * rules; might need work */
		op->o_bd = frontendDB;
	}

	if ( !dryrun ) {
		ID	id;

		if ( be == NULL ) {
			fprintf( stderr, "%s: no target database "
				"has been found for baseDN=\"%s\"; "
				"you may try with \"-u\" (dry run).\n",
				baseDN.bv_val, progname );
			rc = 1;
			goto destroy;
		}

		if ( !be->be_entry_open ||
			!be->be_entry_close ||
			!be->be_dn2id_get ||
			!be->be_entry_get )
		{
			fprintf( stderr, "%s: target database "
				"doesn't support necessary operations; "
				"you may try with \"-u\" (dry run).\n",
				progname );
			rc = 1;
			goto destroy;
		}

		if ( be->be_entry_open( be, 0 ) != 0 ) {
			fprintf( stderr, "%s: could not open database.\n",
				progname );
			rc = 1;
			goto destroy;
		}

		doclose = 1;

		id = be->be_dn2id_get( be, &e.e_nname );
		if ( id == NOID ) {
			fprintf( stderr, "%s: unable to fetch ID of DN \"%s\"\n",
				progname, e.e_nname.bv_val );
			rc = 1;
			goto destroy;
		}
		ep = be->be_entry_get( be, id );
		if ( ep == NULL ) {
			fprintf( stderr, "%s: unable to fetch entry \"%s\" (%lu)\n",
				progname, e.e_nname.bv_val, id );
			rc = 1;
			goto destroy;

		}

		if ( argc == 0 ) {
			Attribute	*a;

			(void)print_access( op, ep, slap_schema.si_ad_entry, NULL, NULL );
			(void)print_access( op, ep, slap_schema.si_ad_children, NULL, NULL );

			for ( a = ep->e_attrs; a; a = a->a_next ) {
				int	i;

				for ( i = 0; !BER_BVISNULL( &a->a_nvals[ i ] ); i++ ) {
					(void)print_access( op, ep, a->a_desc,
							&a->a_vals[ i ],
							&a->a_nvals[ i ] );
				}
			}
		}
	}

	for ( ; argc--; argv++ ) {
		slap_mask_t		mask;
		AttributeDescription	*desc = NULL;
		struct berval		val = BER_BVNULL,
					*valp = NULL;
		const char		*text;
		char			accessmaskbuf[ACCESSMASK_MAXLEN];
		char			*accessstr;
		slap_access_t		access = ACL_AUTH;

		if ( attr == NULL ) {
			attr = argv[ 0 ];
		}

		val.bv_val = strchr( attr, ':' );
		if ( val.bv_val != NULL ) {
			val.bv_val[0] = '\0';
			val.bv_val++;
			val.bv_len = strlen( val.bv_val );
			valp = &val;
		}

		accessstr = strchr( attr, '/' );
		if ( accessstr != NULL ) {
			int	invalid = 0;

			accessstr[0] = '\0';
			accessstr++;
			access = str2access( accessstr );
			switch ( access ) {
			case ACL_INVALID_ACCESS:
				fprintf( stderr, "unknown access \"%s\" for attribute \"%s\"\n",
						accessstr, attr );
				invalid = 1;
				break;

			case ACL_NONE:
				fprintf( stderr, "\"none\" not allowed for attribute \"%s\"\n",
						attr );
				invalid = 1;
				break;

			default:
				break;
			}

			if ( invalid ) {
				if ( continuemode ) {
					continue;
				}
				break;
			}
		}

		rc = slap_str2ad( attr, &desc, &text );
		if ( rc != LDAP_SUCCESS ) {
			fprintf( stderr, "slap_str2ad(%s) failed %d (%s)\n",
					attr, rc, ldap_err2string( rc ) );
			if ( continuemode ) {
				continue;
			}
			break;
		}

		rc = access_allowed_mask( op, ep, desc, valp, access,
				NULL, &mask );

		if ( accessstr ) {
			fprintf( stderr, "%s access to %s%s%s: %s\n",
					accessstr,
					desc->ad_cname.bv_val,
					val.bv_val ? "=" : "",
					val.bv_val ? val.bv_val : "",
					rc ? "ALLOWED" : "DENIED" );

		} else {
			fprintf( stderr, "%s%s%s: %s\n",
					desc->ad_cname.bv_val,
					val.bv_val ? "=" : "",
					val.bv_val ? val.bv_val : "",
					accessmask2str( mask, accessmaskbuf, 1 ) );
		}
		rc = 0;
		attr = NULL;
	}

destroy:;
	if ( !BER_BVISNULL( &e.e_name ) ) {
		ber_memfree( e.e_name.bv_val );
	}
	if ( !BER_BVISNULL( &e.e_nname ) ) {
		ber_memfree( e.e_nname.bv_val );
	}
	if ( !dryrun && be ) {
		if ( ep && ep != &e ) {
			be_entry_release_r( op, ep );
		}
		if ( doclose ) {
			be->be_entry_close( be );
		}

		LDAP_STAILQ_FOREACH( bd, &backendDB, be_next ) {
			if ( bd != be ) {
				backend_shutdown( bd );
			}
		}
	}

	if ( slap_tool_destroy())
		rc = EXIT_FAILURE;

	return rc;
}

